src/modm/driver/can/mcp2515_impl.hpp

```diff
@@ -186,124 +178,232 @@ modm::Mcp2515<SPI, CS, INT>::setMode(Can::Mode mode)
 }
 
 // ----------------------------------------------------------------------------
+
 template <typename SPI, typename CS, typename INT>
 bool
 modm::Mcp2515<SPI, CS, INT>::isMessageAvailable()
 {
-	return !interruptPin.read();
+	return rxQueue.isNotEmpty();
+}
+
+template <typename SPI, typename CS, typename INT>
+bool
+modm::Mcp2515<SPI, CS, INT>::getMessage(can::Message& message, uint8_t* /*filter_id*/)
+{
+	if (rxQueue.isEmpty())
+	{
+		// no message in the receive buffer
+		return false;
+	}
+	else {
+        auto& rxMessage = rxQueue.get();
+		memcpy(&message, &rxMessage, sizeof(message));
+		rxQueue.pop();
+		return true;
+	}
 }
 
 // ----------------------------------------------------------------------------
 template <typename SPI, typename CS, typename INT>
 bool
-modm::Mcp2515<SPI, CS, INT>::getMessage(can::Message& message)
+modm::Mcp2515<SPI, CS, INT>::isReadyToSend()
+{
+	return txQueue.isNotFull();
+}
+
+// ----------------------------------------------------------------------------
+template <typename SPI, typename CS, typename INT>
+bool
+modm::Mcp2515<SPI, CS, INT>::sendMessage(const can::Message& message)
+{
+	bool success = true;
+	if (not modm_assert_continue_ignore(txQueue.push(message), "mcp2515.can.tx",
+			"CAN transmit software buffer overflowed!", 1)) {
+		/// buffer full, could not send
+		success = false;
+	}
+	return success;
+}
+
+// ----------------------------------------------------------------------------
+template <typename SPI, typename CS, typename INT>
+modm::ResumableResult<bool>
+modm::Mcp2515<SPI, CS, INT>::mcp2515ReadMessage()
 {
 	using namespace mcp2515;
 
-	uint8_t status = readStatus(RX_STATUS);
-	uint8_t address;
-	if (status & FLAG_RXB0_FULL) {
-		address = READ_RX;			// message in buffer 0
+	RF_BEGIN();
+
+	// read status flag of the device
+	statusBufferR_ = RF_CALL(readStatus(RX_STATUS));
+
+	tempR_ = true;
```

Comment:
What about a meaningful variable name?

Reply:
well its now called `readTemp` ... thats probably not anymore helpful than before, but done :D